### PR TITLE
Make sure all 301 url redirects get generated when importing more the…

### LINKED ISSUE #32657: [Issue] Make sure all 301 url redirects get generated when importing more the…
This issue is automatically created based on existing pull request: magento/magento2#32640: Make sure all 301 url redirects get generated when importing more the…

---------
…n one product at the same time while using the 'save_rewrites_history' flag.

### Description (*)
Importing multiple products at once while changing their `url_key` and try to generate a 301 redirects (due to setting the `save_rewrites_history` flag) does not work properly, it only generated the 301 redirects for the last product imported.

Issue was introduced in Magento 2.3.4 by https://github.com/magento/magento2/commit/af8d6f4736cb38d0ad619ad1e353700797063b4a#r38859984 (part of https://github.com/magento/magento2/commit/4db926883058842dd67cb1bc28e17d0931fbcef9), I'm guessing the static tests complained about using `array_merge` in a loop, and the author didn't really care about it and replace it with the `+` operator, which is probably equally as bad for performance, but also [works slightly different then array_merge](https://www.texelate.co.uk/blog/combining-arrays-using-plus-versus-array-merge-in-php) which then introduced this bug.

This replaces https://github.com/magento/magento2/pull/29255 and https://github.com/magento/magento2/pull/28099 (this last one wasn't good, because it wasn't able to generate 301 redirects on multiple storeviews, it only fixed importing multiple products, not the combination of multiple products AND multiple storeviews)

### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/magento2/issues/29520

### Manual testing scenarios (*)
1. Create 3 products in the backend of Magento, use these values:

**name**|**sku**|**url\_key**
-----|-----|-----
Product 1|product 1|product-1
Product 2|product 2|product-2
Product 3|product 3|product-3

2. Create the following csv file:

```
sku,store_view_code,product_websites,url_key,save_rewrites_history
product 1,,base,product-1-updated,1
product 2,,base,product-2-updated,1
product 3,,base,product-3-updated,1
```

3. In the backend, go to System > Data Transfer > Import
4. Use Entity Type => Products, Import Behavior => Add/Update and leave the rest as is
5. Import the csv file

### Expected
The `url_rewrite` database table contains 3 `301` redirect types for the 3 products

### Actual
The `url_rewrite` database table contains only 1 `301` redirect type, only for the last imported product

### Questions or comments
I have no interest in adding automated tests for this PR, if somebody else has time/energy/experience, please go ahead!

Should the static tests also search for usages of the `+` operator when merging arrays in a loop and also complain about these use cases?

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by hostep**
…n one product at the same time while using the 'save_rewrites_history' flag.

### Description (*)
Importing multiple products at once while changing their `url_key` and try to generate a 301 redirects (due to setting the `save_rewrites_history` flag) does not work properly, it only generated the 301 redirects for the last product imported.

Issue was introduced in Magento 2.3.4 by https://github.com/magento/magento2/commit/af8d6f4736cb38d0ad619ad1e353700797063b4a#r38859984 (part of https://github.com/magento/magento2/commit/4db926883058842dd67cb1bc28e17d0931fbcef9), I'm guessing the static tests complained about using `array_merge` in a loop, and the author didn't really care about it and replace it with the `+` operator, which is probably equally as bad for performance, but also [works slightly different then array_merge](https://www.texelate.co.uk/blog/combining-arrays-using-plus-versus-array-merge-in-php) which then introduced this bug.

This replaces https://github.com/magento/magento2/pull/29255 and https://github.com/magento/magento2/pull/28099 (this last one wasn't good, because it wasn't able to generate 301 redirects on multiple storeviews, it only fixed importing multiple products, not the combination of multiple products AND multiple storeviews)

### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/magento2/issues/32657

### Manual testing scenarios (*)
1. Create 3 products in the backend of Magento, use these values:

**name**|**sku**|**url\_key**
-----|-----|-----
Product 1|product 1|product-1
Product 2|product 2|product-2
Product 3|product 3|product-3

2. Create the following csv file:

```
sku,store_view_code,product_websites,url_key,save_rewrites_history
product 1,,base,product-1-updated,1
product 2,,base,product-2-updated,1
product 3,,base,product-3-updated,1
```

3. In the backend, go to System > Data Transfer > Import
4. Use Entity Type => Products, Import Behavior => Add/Update and leave the rest as is
5. Import the csv file

### Expected
The `url_rewrite` database table contains 3 `301` redirect types for the 3 products

### Actual
The `url_rewrite` database table contains only 1 `301` redirect type, only for the last imported product

### Questions or comments
I have no interest in adding automated tests for this PR, if somebody else has time/energy/experience, please go ahead!

Should the static tests also search for usages of the `+` operator when merging arrays in a loop and also complain about these use cases?

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#32657: Make sure all 301 url redirects get generated when importing more the…

### CODE REVIEW

**Comment by hostep** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
I would prefer to create a new issue for this PR.

@magento create issue

File: app/code/Magento/CatalogUrlRewrite/Observer/AfterImportDataObserver.php [hunk 1]
```diff
@@ -475,7 +475,7 @@ private function currentUrlRewritesRegenerate()
             ]
         );
 
-        $urlRewrites = [];
+        $urlRewriteArrays = [];
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Observer/AfterImportDataObserver.php]: @hostep, maybe we should use mergeDataProvider, similar to the following code? What do you think?
https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/UrlRewrite/Model/CompositeUrlFinder.php#L74-L83

Comment (hostep) [reply app/code/Magento/CatalogUrlRewrite/Observer/AfterImportDataObserver.php]: That could work indeed, I'll try it out a bit later

**Comment by hostep** [comment]
@ihor-sviziev: I've implemented it now using the `MergeDataProvider` like you suggested. Tested and it works as expected 👍 

The MergeDataProvider class was already used in this file in [the `generateUrls` method](https://github.com/magento/magento2/blob/1d1e97f41f654bcd5db1f7bfaaff2ee54547225c/app/code/Magento/CatalogUrlRewrite/Observer/AfterImportDataObserver.php#L376-L390).
They use it with first creating an instance in the constructor and then cloning that instance when they need it. They introduced this pattern in https://github.com/magento/magento2/commit/f4fb3fcdb0e2c7c1d018dd6ac87d3b61fe0c6f99 (which is part of https://github.com/magento/magento2/commit/1856c28d59c69ea4e5b04502d195b1e872dccc2c#diff-b6b545cd2bf4cd95d7682e9cd23bbb1cd87ae2e538b37d848b58f61899f85cb9) for performance reasons, so I decided to copy it.

@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by ihor-sviziev** [comment]
@sivaschenko @sidolov @gabrieldagama the priority for this PR should be P1, or P2 as it is the redirect from old to new URLs will not be created, so SEO rating for imported products will be affected

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Review by ihor-sviziev — APPROVED** [review]
Approving this PR, hope someone could help with test coverage before delivering this fix

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8996](https://jira.corp.magento.com/browse/ENGCOM-8996) has been created to process this Pull Request

**Comment by engcom-Oscar** [comment]
:heavy_check_mark: QA passed

Was checked case from Manual testing scenarios 

:x: **Before:**
The url_rewrite database table contains only 1 301 redirect type, only for the last imported product

:heavy_check_mark: **After:**
Screenshot from 2021-04-09 12-29-51

**Comment by engcom-Oscar** [comment]
PR need be covered with automation tests.

**Comment by hostep** [comment]
Would be nice if somebody could pick that up, the automation tests, I don't have enough experience with it nor the time for it, thanks! 🙂

**Comment by ihor-sviziev** [comment]
@engcom-Oscar @engcom-Charlie @engcom-Golf, could you help with the test coverage?

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
As we discussed with @engcom-Foxtrot in Slack - there is some sync issue on Github, the commit https://github.com/hostep/magento2/commit/e792abcb1e18e1ebe998252c832f1b7b63a0e515 isn't showing in the PR. Waiting for response from GitHub.

**Comment by hostep** [comment]
Hmm strange, let me know if I need to do something here, but maybe let's wait a bit to see if shows up eventually.

And thanks for adding a test @engcom-Foxtrot !

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, WebAPI Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, WebAPI Tests

**Comment by ihor-sviziev** [comment]
@sidolov @gabrieldagama @sivaschenko, could you explain what does it mean? These 2 builds are failing but don't want to re-run.

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests looks not related to changes from this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8996](https://jira.corp.magento.com/browse/ENGCOM-8996) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@engcom-Foxtrot, thank you so much for your help with the tests!
Moved back to extended testing, as it was moved to the wrong status by the bot.

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, WebAPI Tests

**Comment by hostep** [comment]
Hi folks

Wondering about the status of this PR, it's been in "accepted" status for more than a month, any reason why the merging hasn't happened yet? Thanks!

**Comment by ihor-sviziev** [comment]
@gabrieldagama any thoughts?

**Pull request merged**